bindings/pyroot/pythonizations/python/ROOT/_pythonization/_uhi.py

```diff
@@ -1,23 +1,296 @@
+    if start < _underflow(self, axis) or stop > (_overflow(self, axis) + 1) or start > stop:
+        raise IndexError(f"Slice indices {start, stop} out of range for axis {axis}")
+    return start, stop
+
+
+def _apply_actions(hist, actions):
+    """Apply rebinning or summing actions to the histogram, returns a new histogram"""
+    if not actions or all(a is None for a in actions):
+        return hist
+
+    if any(isinstance(a, _rebin) for a in actions):
+        rebins = [a.ngroup if isinstance(a, _rebin) else 1 for a in actions]
+        hist = _rebin(rebins)(hist)
+
+    if any(a is _sum for a in actions):
+        sum_axes = tuple(i for i, a in enumerate(actions) if a is _sum)
+        return _sum()(hist, sum_axes)
+
+    if any(a is not None and not isinstance(a, (_rebin, _sum)) for a in actions):
+        raise ValueError("Unsupported action detected in actions")
+
+    return hist
+
+
+def _get_processed_slices(self, index):
+    """Process slices and extract actions for each axis"""
+    processed_slices, actions = [], [None] * self.GetDimension()
+    for i, idx in enumerate(index):
+        if isinstance(idx, slice):
+            processed_slices.append(range(idx.start, idx.stop))
+            actions[i] = idx.step
+        else:
+            processed_slices.append([idx])
+
+    return processed_slices, actions
+
+
+def _get_slice_indices(slices):
+    """Generate all combinations of slice indices"""
+    import numpy as np
+
+    return np.array(np.meshgrid(*slices)).T.reshape(-1, len(slices))
+
+
+def _slice_get(self, index):
+    """Retrieve a slice of the histogram based on the index, returns a new histogram"""
+    processed_slices, actions = _get_processed_slices(self, index)
+    slice_indices = _get_slice_indices(processed_slices)
+    target_hist = self.Clone()
+    target_hist.Reset()
+    for indices in slice_indices:
+        indices = tuple(map(int, indices))
+        target_hist.SetBinContent(*indices, self.GetBinContent(self.GetBin(*indices)))
+    target_hist.SetEntries(
+        target_hist.GetEffectiveEntries()
+        + sum(
+            target_hist.GetBinContent(_underflow(target_hist, axis))
+            + target_hist.GetBinContent(_overflow(target_hist, axis))
+            for axis in range(self.GetDimension())
+        )
+    )
+
+    return _apply_actions(target_hist, actions)
+
+
+def _slice_set(self, index, value):
+    """Set values for a slice of the histogram"""
+    import numpy as np
+
+    processed_slices, actions = _get_processed_slices(self, index)
+    slice_indices = _get_slice_indices(processed_slices)
+    if isinstance(value, np.ndarray):
+        if value.size != len(slice_indices):
+            raise ValueError(f"Shape mismatch: expected {len(slice_indices)} values, got {value.size}")
+
+        for indices, val in zip(slice_indices, value.ravel()):
+            _setbin(self, self.GetBin(*map(int, indices)), val)
```

Comment:
I don't know if it's worth it, but we could also think about asserting on the shape of the input array